opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,505 @@
+            static const int contiSolventEqIdx = BlackoilIndices::contiSolventEqIdx;
+            static const int contiPolymerEqIdx = BlackoilIndices::contiPolymerEqIdx;
+
+
+            AquiferCarterTracy(const std::vector<int>& cell_id)
+            : phi_aq_ (1.0), //
+              C_t_ (1.0), //
+              r_o_ (1.0), //
+              k_a_ (1.0), //
+              c1_ (1.0),
+              h_ (1.0), //
+              theta_ (1.0), //
+              c2_ (1.0), //
+              d0_ (1.0),
+              cell_idx_ (cell_id)
+            {
+                mu_w_ = 1e-3;
+                aqutab_td_.push_back(1.0);
+                aqutab_pi_.push_back(1.0);
+                aquiferID_ = 1;
+                inftableID_ = 1;
+                pvttableID_ = 1;
+                init_quantities();
+            }
+
+            explicit AquiferCarterTracy( const AquiferCT::AQUCT_data& params, const Aquancon::AquanconOutput& connection,
+                                         const int numComponents, const Scalar gravity, const Simulator& ebosSimulator    )
+            : phi_aq_ (params.phi_aq), //
+              C_t_ (params.C_t), //
+              r_o_ (params.r_o), //
+              k_a_ (params.k_a), //
+              c1_ (params.c1),
+              h_ (params.h), //
+              theta_ (params.theta), //
+              c2_ (params.c2), //
+              d0_ (params.d0),
+              aqutab_td_ (params.td),
+              aqutab_pi_ (params.pi),
+              aquiferID_ (params.aquiferID),
+              inftableID_ (params.inftableID),
+              pvttableID_ (params.pvttableID),
+              num_components_ (numComponents),
+              gravity_ (gravity),
+              ebos_simulator_ (ebosSimulator)
+            {
+                init_quantities(connection);
+            }
+
+            inline const PhaseUsage&
+            phaseUsage() const
+            {
+                assert(phase_usage_);
+
+                return *phase_usage_;
+            }
+
+            inline int
+            flowPhaseToEbosCompIdx( const int phaseIdx ) const
+            {
+                const auto& pu = phaseUsage();
+                std::cout << "flow 1" << std::endl;
+                if (pu.phase_pos[Water] == phaseIdx)
+                    return BlackoilIndices::canonicalToActiveComponentIndex(FluidSystem::waterCompIdx);
+                if (pu.phase_pos[Oil] == phaseIdx)
+                    return BlackoilIndices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
+                if (pu.phase_pos[Gas] == phaseIdx)
+                    return BlackoilIndices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
+                std::cout << "flow 2" << std::endl;
+                // for other phases return the index
+                return phaseIdx;
+            }
+
+            inline int
+            flowPhaseToEbosPhaseIdx( const int phaseIdx ) const
```

Comment:
I think this function was already removed from the BlackoilModelEbos. I am not good about these index conversion functions, @totto82 or @andlaus can give more insight.

Reply:
Yes. The code should use indices from opm-material and ewoms directly. The only part of the code that still use legacy indices is related to the well code.

Reply:
Ah nice! I was wondering about so much of those boilerplate codes... All set then. Will remove it in the cleanup commits. 👍

Reply:
There are quiet some things are defined while not used. For example, for this model, you do not introduce new primary variables or extra equations.  Some cleaning up of unused stuff will be very helpful.